File: packages/android/src/main/java/com/kozmos/components/RoutingInputGroup/RoutingInputGroup.kt
package com.kozmos.components.routinginputgroup

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsFocusedAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.Place
import androidx.compose.material.icons.filled.SwapVert
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import com.kozmos.providers.KozmosAnalyticsEvent
import com.kozmos.providers.LocalKozmosAnalytics
import com.kozmos.tokens.KozmosDimensions
import com.kozmos.tokens.KozmosThemeTokens
import com.kozmos.components.surface.KozmosSurfaceDefaults
import com.kozmos.components.surface.KozmosSurfaceStyle

data class KozmosRoutePoint(
    val id: String,
    val value: String,
    val placeholder: String? = null
)

/**
 * [surface] is what the group is made of, as React's `surface` prop:
 * [KozmosSurfaceStyle.Solid] (the default) or [KozmosSurfaceStyle.Glass], for a
 * card over the map. Until 2026-09-22 Compose drew it solid only.
 */
@Composable
fun KozmosRoutingInputGroup(
    points: List<KozmosRoutePoint>,
    onPointChange: (String, String) -> Unit,
    modifier: Modifier = Modifier,
    surface: KozmosSurfaceStyle = KozmosSurfaceStyle.Solid,
    onSwap: (() -> Unit)? = null,
    onAddPoint: (() -> Unit)? = null,
    onRemovePoint: ((String) -> Unit)? = null
) {
    val trackEvent = LocalKozmosAnalytics.current

    Surface(
        modifier = modifier,
        shape = RoundedCornerShape(KozmosDimensions.semanticsRadiusPanel),
        // The solid surface React's card sits on by default, themed: the
        // background with the subtle border. It was the background at 90 %
        // under a near-black hairline at 8 % until 2026-09-22.
        color = KozmosSurfaceDefaults.tint(surface),
        tonalElevation = 6.dp,
        shadowElevation = 12.dp,
        border = KozmosSurfaceDefaults.border(surface)
    ) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(KozmosDimensions.primitivesLayoutSpacing200),
            horizontalArrangement = Arrangement.spacedBy(KozmosDimensions.primitivesLayoutSpacing150),
            verticalAlignment = Alignment.Top
        ) {
            RouteTimeline(points)

            Column(
                modifier = Modifier.weight(1f),
                verticalArrangement = Arrangement.spacedBy(KozmosDimensions.primitivesLayoutSpacing150)
            ) {
                points.forEachIndexed { index, point ->
                    Row(
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.spacedBy(KozmosDimensions.primitivesLayoutSpacing100)
                    ) {
                        RoutingField(
                            value = point.value,
                            onValueChange = { onPointChange(point.id, it) },
                            placeholder = point.placeholder ?: defaultPlaceholder(index),
                            modifier = Modifier.weight(1f)
                        )

                        if (points.size > 2 && index > 0 && index < points.lastIndex && onRemovePoint != null) {
                            RoutingAction(
                                icon = Icons.Default.Close,
                                label = "Remove ${point.placeholder ?: point.value.ifEmpty { "route point" }}",
                                filled = false
                            ) {
                                trackEvent(KozmosAnalyticsEvent(component = "RoutingInputGroup", eventName = "point_removed", properties = mapOf("pointId" to point.id)))
                                onRemovePoint(point.id)
                            }
                        }
                    }
                }
            }

            // React's action column: with two points the swap sits 24 down,
            // between the fields, in the secondary fill, and the add 20 below
            // it (8 apart and its own 20); with more, the add is at the top.
            // They were 48 Material icon buttons with no fill until 2026-09-22.
            Column {
                val swaps = points.size == 2 && onSwap != null
                if (swaps) {
                    RoutingAction(
                        icon = Icons.Default.SwapVert,
                        label = "Swap route points",
                        filled = true,
                        modifier = Modifier.padding(top = 24.dp)
                    ) {
                        trackEvent(KozmosAnalyticsEvent(component = "RoutingInputGroup", eventName = "points_swapped"))
                        onSwap?.invoke()
                    }
                }

                if (onAddPoint != null) {
                    RoutingAction(
                        icon = Icons.Default.Add,
                        label = "Add route point",
                        filled = false,
                        modifier = Modifier.padding(top = if (points.size == 2) 20.dp + (if (swaps) 8.dp else 0.dp) else 0.dp)
                    ) {
                        trackEvent(KozmosAnalyticsEvent(component = "RoutingInputGroup", eventName = "point_added"))
                        onAddPoint()
                    }
                }
            }
        }
    }
}

/**
 * React's rail: 12 down, a 14 ring for each point but the last — the start's
 * in the accent over a fifth of it, a waypoint's in the muted foreground — a
 * 2 × 36 connector in the border role between each, and a 16 pin for the end,
 * 8 apart. It was theme/500 with an 18 pin, touching, and its connectors
 * foreground/300 until 2026-09-22.
 */
@Composable
private fun RouteTimeline(points: List<KozmosRoutePoint>) {
    val accent = KozmosThemeTokens.primitivesColorsTheme600
    Column(
        modifier = Modifier.padding(top = 12.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        points.forEachIndexed { index, _ ->
            if (index == points.lastIndex) {
                Icon(
                    imageVector = Icons.Default.Place,
                    contentDescription = null,
                    tint = accent,
                    modifier = Modifier.size(16.dp)
                )
            } else {
                Box(
                    modifier = Modifier
                        .size(14.dp)
                        .background(if (index == 0) accent.copy(alpha = 0.2f) else Color.Transparent, CircleShape)
                        .border(2.dp, if (index == 0) accent else KozmosThemeTokens.primitivesColorsForeground400, CircleShape)
                )
                Box(
                    modifier = Modifier
                        .width(2.dp)
                        .height(36.dp)
                        .background(KozmosThemeTokens.semanticsBorderSubtle, CircleShape)
                )
            }
        }
    }
}

/**
 * One of the group's 40 icon actions, with the control radius. `filled` is the
 * swap's secondary fill with the page ink; the others are ghost in the muted
 * foreground, as React's are.
 */
@Composable
private fun RoutingAction(
    icon: ImageVector,
    label: String,
    filled: Boolean,
    modifier: Modifier = Modifier,
    onClick: () -> Unit
) {
    val shape = RoundedCornerShape(KozmosDimensions.semanticsRadiusControl)
    Box(
        contentAlignment = Alignment.Center,
        modifier = modifier
            .size(40.dp)
            .clip(shape)
            .background(if (filled) KozmosThemeTokens.primitivesColorsBackground200 else Color.Transparent, shape)
            .clickable(role = Role.Button, onClickLabel = label, onClick = onClick)
            .semantics { contentDescription = label }
    ) {
        Icon(
            imageVector = icon,
            contentDescription = null,
            tint = if (filled) KozmosThemeTokens.primitivesColorsForeground0 else KozmosThemeTokens.primitivesColorsForeground400,
            modifier = Modifier.size(16.dp)
        )
    }
}

private fun defaultPlaceholder(index: Int): String {
    return if (index == 0) "Choose Starting Point" else "Choose Destination"
}

/**
 * One route point's field, as React's: 40 high, the control radius, 12 in, no
 * edge at rest but a wash — black at 5 % on light, white at 10 % on dark,
 * doubled while focused — and the ring role's 2 dp outline marking focus.
 * Until 2026-09-22 it was Material's outlined field: 56 high, React's 40, with
 * the dark wash at 5 % and the focus edge in theme/500.
 */
@Composable
private fun RoutingField(
    value: String,
    onValueChange: (String) -> Unit,
    placeholder: String,
    modifier: Modifier = Modifier
) {
    val shape = RoundedCornerShape(KozmosDimensions.semanticsRadiusControl)
    val interaction = remember { MutableInteractionSource() }
    val focused by interaction.collectIsFocusedAsState()
    val wash = (if (KozmosThemeTokens.isDark) 0.10f else 0.05f) * (if (focused) 2f else 1f)
    val ring = KozmosThemeTokens.primitivesColorsTheme600
    val textStyle = MaterialTheme.typography.bodyMedium
    val placeholderColor = KozmosThemeTokens.primitivesColorsForeground400
    BasicTextField(
        value = value,
        onValueChange = onValueChange,
        singleLine = true,
        textStyle = textStyle.copy(color = KozmosThemeTokens.primitivesColorsForeground0),
        cursorBrush = SolidColor(ring),
        interactionSource = interaction,
        modifier = modifier
            .height(40.dp)
            .background(KozmosThemeTokens.primitivesColorsForeground0.copy(alpha = wash), shape)
            .then(if (focused) Modifier.border(2.dp, ring, shape) else Modifier)
            .semantics { contentDescription = placeholder },
        decorationBox = { inner ->
            Box(
                contentAlignment = Alignment.CenterStart,
                modifier = Modifier.fillMaxSize().padding(horizontal = 12.dp)
            ) {
                if (value.isEmpty()) {
                    Text(placeholder, style = textStyle, color = placeholderColor, maxLines = 1)
                }
                inner()
            }
        }
    )
}

File: packages/ios/Sources/Components/RoutingInputGroup/RoutingInputGroup.swift
import SwiftUI

public struct KozmosRoutePoint: Identifiable, Equatable {
    public let id: String
    public let value: String
    public let placeholder: String?

    public init(id: String, value: String, placeholder: String? = nil) {
        self.id = id
        self.value = value
        self.placeholder = placeholder
    }
}

public struct KozmosRoutingInputGroup: View {
    @Environment(\.kozmosAnalytics) private var trackEvent

    private let points: [KozmosRoutePoint]
    private let surface: KozmosSurfaceStyle
    private let onPointChange: (String, String) -> Void
    private let onSwap: (() -> Void)?
    private let onAddPoint: (() -> Void)?
    private let onRemovePoint: ((String) -> Void)?

    /// `surface` is what the group is made of, as React's `surface` prop: `.solid`
    /// (the default) or `.glass`, for a card over the map. Until 2026-09-22
    /// SwiftUI drew it solid only.
    public init(
        points: [KozmosRoutePoint],
        surface: KozmosSurfaceStyle = .solid,
        onPointChange: @escaping (String, String) -> Void,
        onSwap: (() -> Void)? = nil,
        onAddPoint: (() -> Void)? = nil,
        onRemovePoint: ((String) -> Void)? = nil
    ) {
        self.points = points
        self.surface = surface
        self.onPointChange = onPointChange
        self.onSwap = onSwap
        self.onAddPoint = onAddPoint
        self.onRemovePoint = onRemovePoint
    }

    public var body: some View {
        HStack(alignment: .top, spacing: KozmosDimensions.primitivesLayoutSpacing150) {
            routeTimeline

            VStack(spacing: KozmosDimensions.primitivesLayoutSpacing150) {
                ForEach(Array(points.enumerated()), id: \.element.id) { index, point in
                    HStack(spacing: KozmosDimensions.primitivesLayoutSpacing100) {
                        KozmosWashedField(
                            text: Binding(
                                get: { point.value },
                                set: { onPointChange(point.id, $0) }
                            ),
                            placeholder: point.placeholder ?? defaultPlaceholder(for: index)
                        )

                        if canRemove(index: index), let onRemovePoint {
                            iconAction(
                                systemName: "xmark",
                                label: "Remove \(point.placeholder ?? (point.value.isEmpty ? "route point" : point.value))",
                                filled: false
                            ) {
                                trackEvent(KozmosAnalyticsEvent(eventName: "point_removed", component: "RoutingInputGroup", properties: ["pointId": point.id]))
                                onRemovePoint(point.id)
                            }
                        }
                    }
                }
            }

            // React's action column: with two points the swap sits 24 down,
            // between the fields, in the secondary fill, and the add 20 below
            // it (8 apart and its own 20); with more, the add is at the top.
            // They were 40 circles washed at 5 % in foreground/900 until
            // 2026-09-22, the swap 28 down.
            VStack(spacing: 0) {
                let swaps = points.count == 2 && onSwap != nil
                if swaps, let onSwap {
                    iconAction(systemName: "arrow.up.arrow.down", label: "Swap route points", filled: true) {
                        trackEvent(KozmosAnalyticsEvent(eventName: "points_swapped", component: "RoutingInputGroup"))
                        onSwap()
                    }
                    .padding(.top, 24)
                }

                if let onAddPoint {
                    iconAction(systemName: "plus", label: "Add route point", filled: false) {
                        trackEvent(KozmosAnalyticsEvent(eventName: "point_added", component: "RoutingInputGroup"))
                        onAddPoint()
                    }
                    .padding(.top, points.count == 2 ? (swaps ? 28 : 20) : 0)
                }
            }
        }
        .padding(KozmosDimensions.primitivesLayoutSpacing200)
        .clipShape(RoundedRectangle(cornerRadius: KozmosDimensions.semanticsRadiusPanel, style: .continuous))
        // The solid surface React's card sits on by default: the background
        // with the subtle border. It was the background at 90 % under a
        // near-black hairline at 8 % until 2026-09-22.
        .kozmosSurface(RoundedRectangle(cornerRadius: KozmosDimensions.semanticsRadiusPanel, style: .continuous), style: surface)
        .kozmosElevation(KozmosShadows.semanticsElevationOverlay)
    }

    /// React's rail: 12 down, a 14 ring for each point but the last — the
    /// start's in the accent over a fifth of it, a waypoint's in the muted
    /// foreground — a 2 × 36 connector in the border role between each, and a
    /// 16 pin for the end, 8 apart. It was theme/500, 6 apart, its connectors
    /// foreground/300, until 2026-09-22.
    private var routeTimeline: some View {
        VStack(spacing: 8) {
            ForEach(Array(points.enumerated()), id: \.element.id) { index, _ in
                if index == points.count - 1 {
                    Image(systemName: "mappin")
                        .font(.system(size: 14, weight: .semibold))
                        .foregroundColor(KozmosColors.primitivesColorsTheme600)
                        .frame(width: 16, height: 16)
                } else {
                    Circle()
                        .strokeBorder(index == 0 ? KozmosColors.primitivesColorsTheme600 : KozmosColors.primitivesColorsForeground400, lineWidth: 2)
                        .background(
                            Circle()
                                .fill(index == 0 ? KozmosColors.primitivesColorsTheme600.opacity(0.2) : Color.clear)
                        )
                        .frame(width: 14, height: 14)
                    Capsule()
                        .fill(KozmosColors.semanticsBorderSubtle)
                        .frame(width: 2, height: 36)
                }
            }
        }
        .padding(.top, 12)
        .accessibilityHidden(true)
    }

    /// One of the group's 40 icon actions, with the control radius. `filled`
    /// is the swap's secondary fill with the page ink; the others are ghost in
    /// the muted foreground, as React's are.
    private func iconAction(systemName: String, label: String, filled: Bool, action: @escaping () -> Void) -> some View {
        Button(action: action) {
            Image(systemName: systemName)
                .font(.system(size: 13, weight: .semibold))
                .frame(width: 16, height: 16)
                .foregroundColor(filled ? KozmosColors.primitivesColorsForeground0 : KozmosColors.primitivesColorsForeground400)
                .frame(width: 40, height: 40)
                .background(
                    RoundedRectangle(cornerRadius: KozmosDimensions.semanticsRadiusControl, style: .continuous)
                        .fill(filled ? KozmosColors.primitivesColorsBackground200 : Color.clear)
                )
                .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .accessibilityLabel(label)
    }

    private func defaultPlaceholder(for index: Int) -> String {
        index == 0 ? "Choose Starting Point" : "Choose Destination"
    }

    private func canRemove(index: Int) -> Bool {
        points.count > 2 && index > 0 && index < points.count - 1
    }
}
